### ISSUE TITLE
Allow an alternative way of resolving types

### ISSUE THREAD

**Issue by edeustace** [salat/salat#94]
Hi, 
Whilst type hinting is great, it is frustrating that one has to migrate the db each time one re-packages a set of classes that make use of them.

I'd love if I could get Salat to delegate back to my app to resolve the type.

Eg (pseudo code...): 

```
trait TypeResolver{
    def resolveType( collection : String, object : DBObject ) : Class
}

val fooResolver = new TypeResolver{
    def resolveType( collection : String, object : DBObject ) : Class = {
          if( object.get("bar") ) org.foo.Bar else org.foo.Foo
    }
}

context.registerTypeResolver(fooResolver)
```

Bit of a braindump, but just throwing it out there :)

**Comment by noahlz** [comment]
What would the serialized data look like? I assume the goal of this is to simplify migrations when you rename packages / classes etc.

I normally get around this by not having type hints enabled, and using case object overrides.

**Comment by edeustace** [comment]
Yes it's to simplify migrations. Do you have a link to how one would use `registerCaseObjectOverride`?

**Comment by noahlz** [comment]
Hey -  sorry for the delayed response:

https://github.com/salat/salat/blob/1.10.x-branch/salat-core/src/test/scala/salat/test/json/JsonSpec.scala#L198

Another example in my Salat Examples project:
https://github.com/noahlz/salat-examples/commit/3ba02d9fe566e0876c3ff62f010878b291fdecd5

``` scala
scala> SalatExamples.caseObjectOverride()
colors: {
  "_typeHint":"com.novus.salat.examples.Pallet",
  "paint":["red","yellow","blue"]
}

toPrettyJSONArray:
[{
  "_typeHint":"com.novus.salat.examples.Red$"
},{
  "_typeHint":"com.novus.salat.examples.Yellow$"
},{
  "_typeHint":"com.novus.salat.examples.Blue$"
}]
```

Also: TIL that the grater.toJSONArray function doesn't respect case object overrides, and always serializes the type hinted version.  So there's that.
